frontend/src/main/scala/bloop/cli/validation/Validate.scala

```diff
@@ -0,0 +1,85 @@
+package bloop.cli.validation
+
+import java.nio.charset.Charset
+import java.nio.file.{Files, Path}
+
+import bloop.bsp.BspServer
+import bloop.cli.{BspProtocol, Commands, CommonOptions, ExitStatus}
+import bloop.engine.{Action, Exit, Print, Run}
+
+object Validate {
+  private def cliError(msg: String, commonOptions: CommonOptions): Action =
+    Print(msg, commonOptions, Exit(ExitStatus.InvalidCommandLineOption))
+
+  // https://github.com/scalacenter/bloop/issues/196
+  private final val DefaultCharset = Charset.defaultCharset()
+  private[bloop] def bytesOf(s: String): Int = s.getBytes(DefaultCharset).length
+  private final val PipeName = "^\\Q\\\\.\\pipe\\\\E(.*)".r
+  def bsp(cmd: Commands.Bsp, isWindows: Boolean): Action = {
+    val cliOptions = cmd.cliOptions
+    val commonOptions = cliOptions.common
+
+    def validateSocket = cmd.socket match {
+      case Some(socket) if Files.exists(socket) =>
+        cliError(Feedback.existingSocketFile(socket), commonOptions)
+      case Some(socket) if !Files.exists(socket.getParent) =>
+        cliError(Feedback.missingParentOfSocket(socket), commonOptions)
+      case Some(socket) if BspServer.isMac && bytesOf(socket.toString) > 104 =>
+        cliError(Feedback.excessiveSocketLengthInMac(socket), commonOptions)
+      case Some(socket) if bytesOf(socket.toString) > 108 =>
+        cliError(Feedback.excessiveSocketLength(socket), commonOptions)
+      case Some(socket) => Run(Commands.UnixLocalBsp(socket, cliOptions))
+      case None => cliError(Feedback.MissingSocket, commonOptions)
+    }
+
+    def validatePipeName = cmd.pipeName match {
+      case Some(PipeName(pipeName)) => Run(Commands.WindowsLocalBsp(pipeName, cliOptions))
+      case Some(wrong) => cliError(Feedback.unexpectedPipeFormat(wrong), commonOptions)
+      case None => cliError(Feedback.MissingPipeName, commonOptions)
+    }
+
+    def validateTcp = {
+      import java.net.{InetAddress, UnknownHostException}
+      def continueValidation(address: InetAddress): Action = cmd.port match {
+        case n if n > 0 && n <= 1023 => cliError(Feedback.reservedPortNumber(n), commonOptions)
+        case n if n > 1023 && n <= 65535 => Run(Commands.TcpBsp(address, n, cliOptions))
+        case invalid => cliError(Feedback.outOfRangePort(invalid), commonOptions)
+      }
+
+      scala.util.control.Exception
+        .catching(classOf[UnknownHostException])
+        .either { continueValidation(InetAddress.getByName(cmd.host)) } match {
+        case Right(action) => action
+        case Left(_) => cliError(Feedback.unknownHostName(cmd.host), commonOptions)
+      }
+    }
+
+    cmd.protocol match {
+      case BspProtocol.Local if isWindows => validatePipeName
+      case BspProtocol.Local => validateSocket
+      case BspProtocol.Tcp => validateTcp
+    }
+  }
+}
+
+object Feedback {
```

Comment:
Would be cool in the future to extract that to another file and use it for all the feedback that Bloop has to offer.